File: Cpp/440.k-th-smallest-in-lexicographical-order.cpp
class Solution {
public:
    long getCount(long prefix, long n) {
        long cur = prefix;
        long next = cur + 1;
        long nodeCount = 0;
        while (cur <= n) {
            nodeCount += min(n+1, next) - cur;
            cur *= 10;
            next *= 10;
        }
    }

    int findKthNumber(int n, int k) {
        long p = 1;
        long prefix = 1;
        while (p < k) {
            long count = getCount(prefix, n);
            if (p + count > k) {
                prefix *= 10;
                p++;
            }
            else {
                prefix++;
                p += count;
            }
        }
        
        return static_cast<int>(prefix);
    }
};

File: Python/440.k-th-smallest-in-lexicographical-order.py
class Solution:

    def findKthNumber(self, n, k) -> int:
        cur = 1
        prefix = 1

        while cur < k:
            cnt = self.get_count(prefix, n)
            if cur + cnt > k:
                prefix *= 10
                cur += 1
            else:
                prefix += 1
                cur += cnt
        return prefix

    def get_count(self, i, n):
        if i <= n:
            cnt = 1
        else:
            return 0

        a = i
        b = i + 1

        while True:
            a = a * 10
            b = b * 10
            if n >= b:
                cnt += b - a
            elif n >= a:
                cnt += n - a + 1
            else:
                break
        return cnt
